packages/drivers/odsp-driver/src/handleReadonlyServerState.ts

```diff
@@ -0,0 +1,56 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert, delay } from "@fluidframework/common-utils";
+import { getRetryDelayFromError } from "@fluidframework/driver-utils";
+import { OdspErrorType } from "@fluidframework/odsp-driver-definitions";
+import { FetchType } from "./epochTracker";
+
+export async function handleReadOnlyServerState<T>(
+    api: () => Promise<T>,
+    logger: ITelemetryLogger,
+    fetchType: FetchType,
+) {
+    let result: T | undefined;
+    let success = false;
+    let retryAfterMs = 1000; // has to be positive!
+    let numRetries = 0;
+    const maxRetries = 5;
+    let lastError: any;
+    do {
+        try {
+            result = await api();
+            success = true;
+        } catch (error) {
+            lastError = error;
+            // If the innerMostError is not "ServiceReadonly" error or we have retried the max times, then just
+            // give up.
+            if (error?.innerMostErrorCode !== OdspErrorType.serviceReadOnly || numRetries >= maxRetries) {
```

Comment:
I do not think this code should work with innerMostErrorCode, instead it should use resolved errorType from createOdspNetworkError() and thus have logic on what is OdspErrorType.serviceReadOnly in one place.

Reply:
Removed this code and merged with storage adapter for retry.